### PR TITLE
Implementation of ImportDirectoryBuildProps and ImportDirectoryBuildTargets

### PR DESCRIPTION
**Pull request by jeffkl**
#222 MSBuild should have a solution import in Microsoft.Common.Targets

Here is my initial implementation of the "directory" import.  I optimized it as much as possible by not continuing if GetDirectoryNameOfFileAbove() returns String.Empty.

1. _DirectoryBuildPropsFile is sort of private with the underscore and is only there so that I didn't have global.build.props in a few places.  I can remove it if we want.
2. _DirectoryBuildPropsBasePath is used to store the value of GetDirectoryNameOfFileAbove() in the event that the file isn't found.
3. DirectoryBuildPropsPath is the property that stores the combination of the file and base path, unless the user overrides it.

Unit tests have also been implemented.

### CODE REVIEW

**Comment by jeffkl** [comment]
@dotnet-bot test this

**Comment by rainersigwald** [comment]
@jeffkl Pretty sure you have to be more polite than that.

(In other words, the "please" is mandatory.  I didn't configure the bot)

**Comment by jeffkl** [comment]
Haha that's awesome.  

@dotnet-bot test this please?

File: src/XMakeTasks/Microsoft.Common.targets [hunk 1]
```diff
@@ -121,7 +121,23 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        Determine the path to the global build targets file if the user did not disable $(ImportGlobalBuildTargets) and
+        they did not already specify an absolute path to use via $(GlobalBuildTargetsPath)
+    -->
+  <PropertyGroup Condition="'$(ImportGlobalBuildTargets)' == 'true' and '$(GlobalBuildTargetsPath)' == ''">
+    <_GlobalBuildTargetsFile Condition="'$(_GlobalBuildTargetsFile)' == ''">global.build.targets</_GlobalBuildTargetsFile>
+    <_GlobalBuildTargetsBasePath Condition="'$(_GlobalBuildTargetsBasePath)' == ''">$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), '$(_GlobalBuildTargetsFile)'))</_GlobalBuildTargetsBasePath>
+    <GlobalBuildTargetsPath Condition="'$(_GlobalBuildTargetsBasePath)' != '' and '$(_GlobalBuildTargetsFile)' != ''">$([System.IO.Path]::Combine('$(_GlobalBuildTargetsBasePath)', '$(_GlobalBuildTargetsFile)'))</GlobalBuildTargetsPath>
+  </PropertyGroup>
+
+  <Import Project="$(GlobalBuildTargetsPath)" Condition="'$(ImportGlobalBuildTargets)' == 'true' and exists('$(GlobalBuildTargetsPath)')"/>
```
Comment (rainersigwald) [inline src/XMakeTasks/Microsoft.Common.targets]: I think it makes sense to put this block before the `ImportAfter` imports. That seems to better go with the intent of "before/after".

Comment (jeffkl) [inline src/XMakeTasks/Microsoft.Common.targets]: I agree.  Do you think it should import before or after Microsoft.Common.CurrentVersion.targets?

**Comment by rainersigwald** [comment]
Odd, the first failure indicated one failed test, `Microsoft.Build.UnitTests.OM.Instance.ProjectInstance_Internal_Tests.Build`. The second looks like xunit crashed without writing a results file.

@dotnet-bot test this please

File: src/XMakeTasks/Microsoft.Common.targets (hunk 1 above)
Comment (jeffkl) [inline src/XMakeTasks/Microsoft.Common.targets]: Ping?  Do you like the new placement?  Is everyone okay with the unit tests?

Comment (rainersigwald) [inline src/XMakeTasks/Microsoft.Common.targets]: I think I'd go after `CommonTargetsPath`. That would make it easier to override targets if you needed to.

---
In reply to: [69346571](https://github.com/Microsoft/msbuild/pull/751#discussion_r69346571) [](ancestors = 69346571)

Comment (jeffkl) [inline src/XMakeTasks/Microsoft.Common.targets]: Done

**Comment by rainersigwald** [comment]
Implementation and tests look good!

I'm not totally settled about the names. "Global" makes sense by analogy to `global.json` in .NET CLI-land, but that's not commonly used now. And these aren't global settings--the existing machine- or user-wide magic imports would be, but these are limited to child directories.

**Comment by jeffkl** [comment]
Hmm I'm afraid because the build keeps failing because of something weird.

@dotnet-bot test this please

**Comment by rainersigwald** [comment]
LGTM

**Comment by AndyGerlicher** [comment]
LGTM

**Pull request merged**